### PR TITLE
Closes #1193: Feature: add synchronized author/post outline in desktop view

### LINKED ISSUE #1193: Feature: add synchronized author/post outline in desktop view
**What would you like to be added**:

In desktop view, I'd like to have a high-level view of all the posts and authors in the current timeline, so I can jump around easily.

Here's a functional example of what I mean from the [Jest Docs](https://jestjs.io/docs/en/mock-function-api):

scroll-sync

On the right, all of the document's major sections are listed, and as I scroll the main page, the list on the right gets synchronized with where I am.  I can easily see where I am in the entire document.

Imagine if instead of that info, we had Post Title and Author Title on the side.  We have more than enough space on the left/right of the posts to add this.

**Why would you like this to be added**:

Moving through the posts linearly is great, and I can read one, then the next, etc.  However, I also want to be able to jump around, or go back to an author or post I read a few minutes ago.  We need better visibility into what is in the feed.

This needs a little bit of design thought, too.

### PR DESCRIPTION
**Pull request by manekenpix**
## Issue This PR Addresses
Closes #1193 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [x] **UI**: Change which improves UI

## Description

This adds an author/post outline in the desktop view.

To keep track of posts entering and leaving the viewport, I used [react-intersection-observer](https://www.npmjs.com/package/react-intersection-observer). I saw the solution used in #1332, but I thought this module provides more functionality and [testing tools](https://www.npmjs.com/package/react-intersection-observer#testing).

For communication between the outline and the timeline components, I used [useContext](https://reactjs.org/docs/hooks-reference.html#usecontext).

Originally, keeping track of posts' headers would've been the best approach to know when a post enters or leaves the viewport, but after all the improvements and the way headers behave now, I had to switch to tracking posts' content. This causes weird behaviours sometimes, but I'm working on solving that.

This is a first iteration that mainly focuses on functionality. Styling can be improved in follow-up pull requests (please don't make me do it).

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
This is so cool, I just tried playing with it on Vercel and love it!  I'll try to prioritize reviewing this soon.  Thank you for doing it @manekenpix.

File: src/frontend/src/components/Posts/Outline/Entry.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+  if (!post) {
+    return <div>Loading...</div>;
+  }
+
+  return (
+    <ListItem
+      button
+      component="a"
+      href={`#inView${post.id}`}
```
Comment (humphd) [inline src/frontend/src/components/Posts/Outline/Entry.jsx:32]: Is there any way that we can clean this up to get rid of the `inView` prefix?  Do we already use the post id asn id elsewhere?

File: src/frontend/src/components/Posts/Outline/Outline.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  if (!posts) return <div>Loading...</div>;
+
+  const postsUrl = posts.map((set) =>
+    set.map(({ id, url }) => <Entry postUrl={`${telescopeUrl}${url}`} key={id} />)
+  );
+
+  return (
+    <ListSubheader className={classes.header} component="nav">
+      {postsUrl && postsUrl.map((post) => post)}
```
Comment (humphd) [inline src/frontend/src/components/Posts/Outline/Outline.jsx:30]: Why bother with `postsUrl.map((post) => post)`?

File: src/frontend/src/components/Posts/Posts.jsx [hunk 3]
```diff
@@ -30,6 +31,7 @@ const REFRESH_INTERVAL = 5 * 60 * 1000; /* refresh data every 5 minutes */
 
 const Posts = () => {
   const classes = useStyles();
+  const [isDesktopView, setDesktopView] = useState(true);
```
Comment (humphd) [inline src/frontend/src/components/Posts/Posts.jsx:34]: We had to do something hacky like this for the sticky header views too.  I wonder if we can abstract this out into a hook or context we (re)use at the top level?

File: src/frontend/src/components/Posts/Posts.jsx [hunk 4]
```diff
@@ -39,6 +41,16 @@ const Posts = () => {
     }
   );
 
+  const updateMedia = () => {
+    setDesktopView(window.innerWidth > 1450);
```
Comment (humphd) [inline src/frontend/src/components/Posts/Posts.jsx:45]: I think we can do this at much narrower views.  We actually have more space if we shift the timeline of posts to the right and drop the right margin.

File: src/frontend/src/components/Posts/Posts.jsx [hunk 5]
```diff
@@ -39,6 +41,16 @@ const Posts = () => {
...
+  };
+
+  useEffect(() => {
+    updateMedia();
+    window.addEventListener('resize', updateMedia);
+    return () => window.removeEventListener('resize', updateMedia);
+  });
```
Comment (humphd) [inline src/frontend/src/components/Posts/Posts.jsx:52]: This should have some kind of dependency so we don't keep running it.

File: src/frontend/src/components/Posts/Posts.jsx [hunk 6]
```diff
@@ -60,7 +72,16 @@ const Posts = () => {
     );
   }
 
-  return (
+  return isDesktopView ? (
```
Comment (humphd) [inline src/frontend/src/components/Posts/Posts.jsx:75]: Instead of these long switches, why not make 2 sub components and render one or the other?

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by PedroFonsecaDEV** [comment]
Following.

**Comment by manekenpix** [comment]
Since #1316 is the priority now, I think it's a good idea to close this one and try #1193 again once the Gatsby-Next port is completed.

**Pull request closed without merging**